src/main/java/com/google/googleidentity/oauth2/token/InMemoryOAuth2TokenService.java

```diff
@@ -169,13 +159,16 @@ private OAuth2AccessToken getNewAccessToken(
 
     String accessTokenValue = UUID.randomUUID().toString();
 
-    while (user.readAccessToken(clientID, accessTokenValue).isPresent()) {
-      accessTokenValue = UUID.randomUUID().toString();
-    }
-
     String accessTokenString =
         new UserClientTokenInfo(username, clientID, accessTokenValue).getEncryptTokenString(aesKey);
 
+    while (user.readAccessToken(clientID, accessTokenString).isPresent()) {
+      accessTokenValue = UUID.randomUUID().toString();
+      accessTokenString =
+          new UserClientTokenInfo(username, clientID, accessTokenValue)
+              .getEncryptTokenString(aesKey);
```

Comment:
I noticed this getEncryptTokenString returns optional, actually if this function failed, the later logic should not work, so instead of return optional, it shall return String, and throw exception when it fails. Same for other critical token generation functions. You may fix this in later pull request, not in this one to avoid too many changes.

Reply:
getEncryptTokenString actually returns string . The exception has been hanlded in getEncryptTokenString itself

Reply:
Hmm, I read it wrong, but the point is still the same, since we don't really tolerate accessTokenString to be null, we'd better just throw out the exception in the first place.